dbms/src/Functions/FunctionsHashing.h

```diff
@@ -353,6 +353,43 @@ struct JavaHashImpl
     static constexpr bool use_int_hash_for_pods = false;
 };
 
+struct JavaHashUTF16LEImpl
+{
+    static constexpr auto name = "javaHashUTF16LE";
+    using ReturnType = Int32;
+
+    static Int32 apply(const char * raw_data, const size_t raw_size)
+    {
+        char * data = const_cast<char *>(raw_data);
+        size_t size = raw_size;
+
+        // Remove Byte-order-mark(0xFFFE) for UTF-16LE
+        if (size >= 2 && data[0] == -1 && data[1] == -2)
+        {
+            data += 2;
+            size -= 2;
+        }
+
+        if (size % 2 != 0)
+            throw Exception("Arguments for javaHashUTF16LE must be in the form of UTF-16", ErrorCodes::LOGICAL_ERROR);
+
+        UInt32 h = 0;
+        for (size_t i = 0; i < size; i += 2)
+            h = 31 * h + static_cast<UInt32>((data[i] & 0xFF) << HI_BYTE_SHIFT | (data[i + 1] & 0xFF) << LO_BYTE_SHIFT);
```

Comment:
This code is difficult to read, because it involves `char` type that is signed on x86 but unsigned on ARM,
and on x86 it involves bit shift of negative number
and cast to UInt32 with overflow.

Have you checked that it works correctly?

PS. char & 0xFF is redundant.
HI_BYTE_SHIFT and LO_BYTE_SHIFT can be inline values.
You can also use `unalignedLoad`